src/java/htsjdk/samtools/SAMRecord.java

```diff
@@ -447,24 +423,18 @@ public String getMateReferenceName() {
      * mate reference index to NO_ALIGNMENT_REFERENCE_INDEX.
      *
      * @param mateReferenceName - must not be null
+     * @throws IllegalArgumentException if {@code mateReferenceName} is null
      */
     public void setMateReferenceName(final String mateReferenceName) {
         if (null == mateReferenceName) {
-            throw new IllegalArgumentException("Mate reference name must not be null");
+            throw new IllegalArgumentException("Mate reference name must not be null. Use SAMRecord.NO_ALIGNMENT_REFERENCE_NAME to reset the mate reference name.");
         }
-        else if (NO_ALIGNMENT_REFERENCE_NAME.equals(mateReferenceName)) {
-            mMateReferenceName = NO_ALIGNMENT_REFERENCE_NAME;
-            mMateReferenceIndex = NO_ALIGNMENT_REFERENCE_INDEX;
-        }
-        else if (null != mHeader) {
-            final int mateReferenceIndex = mHeader.getSequenceIndex(mateReferenceName);
-            if (-1 != mateReferenceIndex) {
-                setMateReferenceIndex(mateReferenceIndex); // sets mate reference name and index
-            }
-            else {
-                mMateReferenceName = mateReferenceName.intern();
-                mMateReferenceIndex = null;
-            }
+        if (null != mHeader) {
+            mMateReferenceIndex = resolveIndexFromName(mateReferenceName, mHeader, false);
+            // String.intern() is surprisingly expensive, so try to avoid it by looking up in sequence dictionary if possible
+            mMateReferenceName = null == mMateReferenceIndex ?
+                                   mateReferenceName.intern() :
+                                   resolveNameFromIndex(mMateReferenceIndex, mHeader);
```

Comment:
Same comment as for `setReferenceName()` -- why do we make this additional call to resolveNameFromIndex() here?